component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java

```diff
@@ -209,9 +257,155 @@ protected boolean authenticate(X509Certificate registeredCert, X509Certificate r
+                    .getValueFromCache(mutualTLSJWKSCacheKey);
+            if (mutualTLSJWKSCacheEntry != null) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Retrieving JWKS for " + jwksUri.toString() + " from cache.");
+                }
+                resource = mutualTLSJWKSCacheEntry.getValue();
+                if (log.isDebugEnabled() && resource != null) {
+                    log.debug("Cache hit for " + jwksUri.toString());
+                }
+            }
+            if (resource == null) {
+
+                DefaultResourceRetriever defaultResourceRetriever;
+                defaultResourceRetriever = new DefaultResourceRetriever(
+                        MutualTLSUtil.readHTTPConnectionConfigValue(HTTP_CONNECTION_TIMEOUT_XPATH),
+                        MutualTLSUtil.readHTTPConnectionConfigValue(HTTP_READ_TIMEOUT_XPATH));
+                if (log.isDebugEnabled()) {
+                    log.debug("Fetching JWKS from remote endpoint. JWKS URI: " + jwksUri);
+                }
+                resource = defaultResourceRetriever.retrieveResource(jwksUri);
+                MutualTLSJWKSCache.getInstance()
+                        .addToCache(mutualTLSJWKSCacheKey, new MutualTLSJWKSCacheEntry(resource));
+            }
+            if (resource != null) {
+                JsonParser jp = new JsonParser();
+                try (InputStream inputStream = new ByteArrayInputStream(
+                        resource.getContent().getBytes(StandardCharsets.UTF_8));
+                        InputStreamReader inputStreamReader = new InputStreamReader(inputStream)) {
+                    JsonElement root = jp.parse(inputStreamReader);
+                    JsonObject rootObj = root.getAsJsonObject();
+                    JsonElement keys = rootObj.get(KEYS);
+                    if (keys != null) {
+                        return keys.getAsJsonArray();
+                    } else {
+                        return null;
+                    }
+                }
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Fetch JWKS endpoint using client ID.
+     *
+     * @param serviceProvider Service Provider
+     */
+    public URL getJWKSEndpointOfSP(ServiceProvider serviceProvider, String clientID) throws OAuthClientAuthnException {
+
+        String jwksUri;
+        jwksUri = MutualTLSUtil.getPropertyValue(serviceProvider, JWKS_URI);
+        if (jwksUri != null) {
+            URL url;
+            try {
+                url = new URL(jwksUri);
+                if (log.isDebugEnabled()) {
+                    log.debug("Configured JWKS URI found: " + jwksUri);
+                }
+            } catch (MalformedURLException e) {
+                throw new OAuthClientAuthnException("URL might be malformed " + clientID, OAuth2ErrorCodes.SERVER_ERROR,
+                        e);
+            }
+            return url;
+        } else {
```

Comment:
Rather than having an else block here we can improve the logic of this method by checking for null at the very beginning as follows. 
```
if (jwksUri == null) {
     throw new OAuthClientAuthnException(
                    "jwks endpoint not configured for the service provider for client ID: " + clientID,
                    OAuth2ErrorCodes.SERVER_ERROR);
}
```

Reply:
Fixed https://github.com/wso2-extensions/identity-oauth-addons/pull/52